kotlinx-coroutines-debug/src/internal/DebugProbesImpl.kt

```diff
@@ -249,32 +314,40 @@ internal object DebugProbesImpl {
          * Here we replace completion with a sequence of CoroutineStackFrame objects
          * which represents creation stacktrace, thus making stacktrace recovery mechanism
          * even more verbose (it will attach coroutine creation stacktrace to all exceptions),
-         * and then using this artificial frame as an identifier of coroutineSuspended/resumed calls.
+         * and then using CoroutineOwner completion as unique identifier of coroutineSuspended/resumed calls.
          */
         val stacktrace = sanitizeStackTrace(Exception())
         val frame = stacktrace.foldRight<StackTraceElement, CoroutineStackFrame?>(null) { frame, acc ->
             object : CoroutineStackFrame {
                 override val callerFrame: CoroutineStackFrame? = acc
                 override fun getStackTraceElement(): StackTraceElement = frame
             }
-        }
-        return ArtificialStackFrame(completion, frame!!).also {
-            storeFrame(it, completion)
-        }
+        }!!
+
+        return createOwner(completion, frame)
     }
 
     @Synchronized
-    private fun <T> storeFrame(frame: ArtificialStackFrame<T>, completion: Continuation<T>) {
-        capturedCoroutines[frame] = CoroutineState(completion, frame, ++sequenceNumber)
+    private fun <T> createOwner(completion: Continuation<T>, frame: CoroutineStackFrame): Continuation<T> {
+        if (!isInstalled) return completion
+        val info = CoroutineInfo(completion.context, frame, ++sequenceNumber)
+        val owner = CoroutineOwner(completion, info, frame)
+        capturedCoroutines += owner
+        return owner
     }
 
     @Synchronized
-    private fun probeCoroutineCompleted(coroutine: ArtificialStackFrame<*>) {
-        capturedCoroutines.remove(coroutine)
+    private fun probeCoroutineCompleted(owner: CoroutineOwner<*>) {
+        capturedCoroutines.remove(owner)
```

Comment:
Need to also remove `owner.info.lastObservedFrame.realCaller()` from `infoCache`.